Make the path of the merged log configurable, with the default being …

vsiakka opened the pull request:
…merged/results.json to keep backwards compatibility

@midniteio @efokschaner

vsiakka committed:
Make the path of the merged log configurable, with the default being merged/results.json to keep backwards compatibility

vsiakka committed:
Fix lint

vsiakka committed:
Update test fixtures to fix unit tests

efokschaner commented on src/lib/cli.js:
```diff
@@ -28,6 +28,7 @@ program
   .option('-d, --devMode', 'Shortcut for running cucumber-js directly')
   .option('--fail-fast', 'abort the run on first failure')
   .option('--strict', 'Fail fast if a step is undefined')
+  .option('--mergedLog', 'Path, relative to logdir, merged log. Specify "" to not generate one.')
```
There's already inconsistency in here but I think we want cmd line params hyphenated, not camelcase. Would it be more natural for paths in command line params to be relative to cwd?

efokschaner replied:
Also mention the default value in the command line help. Bonus points if you actually format the default value in to the help to keep it DRY.

efokschaner commented on test/test-handler-spec.js:
```diff
@@ -68,7 +68,7 @@ describe('Test Handler', function() {
   });
 
-  it('should generate a merged json log', function () {
+  it('should generate a merged json log by default', function () {
```
Maybe add a new test to lock down the empty string behaviour?

efokschaner commented:
Had a few small suggestions but :+1: overall.

efokschaner reviewed (approved):
Had a few small suggestions but 👍 overall.

vsiakka committed:
Make all commandline parameters hyphanated, make mergelog path not be relative to the log dir, add unit test for mergedLog being an empty string, move default options to a seperate file so we can use them from many places

vsiakka commented:
@efokschaner I put up another iteration which I think addresses all your comments. Do you want to take another look? Fyi I will do a major version bump since I am changing the cli options.

efokschaner commented on test/test-handler-spec.js:
```diff
@@ -68,16 +69,35 @@ describe('Test Handler', function() {
+    var opts = defaultOptions;
+    this.timeout(timeoutMs);
+    let mergedFilePath = defaultOptions.mergedLog;
+    opts.mergedLog = "";
```
You're mutating `defaultOptions` here. Maybe that's safe with the way the test framework works, not sure.

efokschaner commented:
Nice cleanup 👍

vsiakka committed:
Clone the options object before changing it

midniteio commented on src/lib/cli.js:
```diff
@@ -18,16 +19,19 @@ program
+  .option('-d, --dev-mode', 'Shortcut for running cucumber-js directly')
   .option('--fail-fast', 'abort the run on first failure')
   .option('--strict', 'Fail fast if a step is undefined')
+  .option('--merged-log <path>', 'Path for merged log. Specify "" to not generate one. ' +
```
It feels like we're trying to overload the path with also handling a 'should-generate-logs' boolean sort of flag by the use of omiting a path here. I think it'd probably be cleaner to provide both the `--merged-log` flag along with something like `--disable-merged-log` or `--no-merged-log`

midniteio commented on src/lib/cli.js:
```diff
@@ -18,16 +19,19 @@ program
     '*Note* This adds complexity to the logs that are hard to decipher, but included if needed for debugging'
   )
-  .option('-d, --devMode', 'Shortcut for running cucumber-js directly')
+  .option('-d, --dev-mode', 'Shortcut for running cucumber-js directly')
```
I'm definitely aligned with consistency for these flags and do think kebab case is the right move here, but removing this backwards compatibility feels unnecessary at the moment. Since this arg parsing isn't tied to a command, I think we could do some smart stuff such as keep these changes above as additions, and also keep the old  in a commented block like:
```
...
.option('--strict', 'Fail fast if a step is undefined')
// Deprecated flags to be removed in v1.0.0
.option('-l, --logdir <path>', 'Output dir for test logs')
.option('-d, --devMode', 'Shortcut for running cucumber-js directly')
(etc) ...
```
And then in the export:
```
export const args = {	
  requires: (program.require.length) ? program.require : undefined,
  cucumberPath: program.cucumber,		 
  workers: program.workers,	
  logdir: program.logdir || program.logDir,
  silentSummary: program.silentsummary || program.silentSummary,
  ...
```

This basically also buys us some deprecation notices, in case anyone is using this in their CI:
```
let deprecations = [
  { oldFlag: '--logdir', newFlag: '--log-dir'},
  { oldFlag: '--devMode', newFlag: '--dev-mode'}
  ...
];
```
_Before the export, after `.parse(process.argv)`_
```
deprecations.forEach(({oldFlag, newFlag}) => {
  if(program.rawArgs.includes(oldFlag)) {
    console.warn(`$.oldFlag} is deprecated and will be removed in future versions. Please use ${newflag} instead`);
  }
});
```

vsiakka committed:
Make cli changes backwards compatible and add --disable-merged-log
